File: src/javascript/algorithms/linkedlist/8.findKthNodeFromEnd/findKthNodeFromEnd.js
class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}

class LinkedList {
    constructor(value) {
        this.head = null;
        this.tail = null;
        this.length = 0;
    }

    push(value) {
        const newNode = new Node(value);
        if(!this.head) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            this.tail.next = newNode;
            this.tail = newNode;
        }
        this.length++;
        return this;
    }

    findKthNodeFromEnd(k) {
        let first = this.head;
        let second = this.head;
        for (let i = 0; i < k; i++) {
            if (first === null) return null;
            first = first.next;
        }

        while(first != null) {
            second = second.next;
            first = first.next;
        }
        return second;
    }
}

const myLinkedList = new LinkedList();
myLinkedList.push(1);
myLinkedList.push(2);
myLinkedList.push(3);
myLinkedList.push(4);
myLinkedList.push(5);

console.dir(myLinkedList.findKthNodeFromEnd(2),{depth: null});
console.dir(myLinkedList.findKthNodeFromEnd(5),{depth: null});

File: src/python/algorithms/linkedlist/8.findKthNodeFromEnd/find_kth_node_from_end.py
"""
Find Kth Node From End

Given the head of a linked list and an integer k,
return the kth node from the end of the list.

Time Complexity: O(n)
Space Complexity: O(1)
"""


class ListNode:
    def __init__(self, value=0):
        self.value = value
        self.next = None


def find_kth_node_from_end(head, k):
    """
    Two pointer approach with k distance gap.
    TC: O(n), SC: O(1)
    """
    if not head or k <= 0:
        return None
    
    slow = head
    fast = head
    
    # Move fast pointer k steps ahead
    for _ in range(k):
        if not fast:
            return None  # k is larger than list length
        fast = fast.next
    
    # Move both pointers until fast reaches end
    while fast:
        slow = slow.next
        fast = fast.next
    
    return slow


def print_list(head):
    """Helper to print linked list."""
    result = []
    while head:
        result.append(str(head.value))
        head = head.next
    return " -> ".join(result)


# Test cases
if __name__ == "__main__":
    # Create list: 1 -> 2 -> 3 -> 4 -> 5
    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(3)
    head.next.next.next = ListNode(4)
    head.next.next.next.next = ListNode(5)
    
    print(f"List: {print_list(head)}")
    
    k = 2
    result = find_kth_node_from_end(head, k)
    print(f"{k}th node from end: {result.value}")  # 4
    
    k = 1
    result = find_kth_node_from_end(head, k)
    print(f"{k}st node from end: {result.value}")  # 5
    
    k = 5
    result = find_kth_node_from_end(head, k)
    print(f"{k}th node from end: {result.value}")  # 1
    
    k = 6
    result = find_kth_node_from_end(head, k)
    print(f"{k}th node from end: {result}")  # None
